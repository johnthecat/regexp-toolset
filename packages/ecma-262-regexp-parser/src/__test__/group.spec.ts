import { expect, describe, it } from 'vitest';
import { parseRegexp } from '../index.js';

describe('Group', () => {
  describe('Common', () => {
    it('should throw, if group is not closed', () => {
      expect(() => parseRegexp('/(ab/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /(ab/
            ════
         Incomplete group structure"
      `);
    });

    it('should throw, there is unmatched parenthesis', () => {
      expect(() => parseRegexp('/(ab))/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /(ab))/
                ═
         Unmatched parenthesis"
      `);
    });

    it('should correctly parse nested groups', () => {
      expect(parseRegexp('/(a(f)b)/')).toMatchSnapshot();
    });
  });

  describe('Capturing group', () => {
    it('should parse simple capturing group', () => {
      expect(parseRegexp('/(ab)/')).toMatchSnapshot();
    });
  });

  describe('Non Capturing group', () => {
    it('should parse simple non capturing group', () => {
      expect(parseRegexp('/(?:ab)/')).toMatchSnapshot();
    });
  });

  describe('Positive lookahead', () => {
    it('should parse simple positive lookahead', () => {
      expect(parseRegexp('/(?=ab)/')).toMatchSnapshot();
    });
  });

  describe('Negative lookahead', () => {
    it('should parse simple negative lookahead', () => {
      expect(parseRegexp('/(?!ab)/')).toMatchSnapshot();
    });
  });

  describe('Positive lookbehind', () => {
    it('should parse simple positive lookbehind', () => {
      expect(parseRegexp('/(?<=ab)/')).toMatchSnapshot();
    });
  });

  describe('Negative lookbehind', () => {
    it('should parse simple negative lookbehind', () => {
      expect(parseRegexp('/(?<!ab)/')).toMatchSnapshot();
    });
  });
});
