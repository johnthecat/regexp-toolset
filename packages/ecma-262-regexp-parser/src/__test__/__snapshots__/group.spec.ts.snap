// Vitest Snapshot v1

exports[`Group > Capturing group > should parse simple capturing group 1`] = `
{
  "body": {
    "body": {
      "end": 3,
      "expressions": [
        {
          "charCode": 97,
          "end": 2,
          "kind": 6,
          "start": 2,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 3,
          "kind": 6,
          "start": 3,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 2,
    },
    "end": 4,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "capturing",
  },
  "end": 5,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Common > should correctly parse nested groups 1`] = `
{
  "body": {
    "body": {
      "end": 8,
      "expressions": [
        {
          "charCode": 97,
          "end": 2,
          "kind": 6,
          "start": 2,
          "type": "simple",
          "value": "a",
        },
        {
          "body": {
            "end": 6,
            "kind": 3,
            "left": {
              "charCode": 102,
              "end": 4,
              "kind": 6,
              "start": 4,
              "type": "simple",
              "value": "f",
            },
            "right": {
              "charCode": 103,
              "end": 6,
              "kind": 6,
              "start": 6,
              "type": "simple",
              "value": "g",
            },
            "start": 4,
          },
          "end": 7,
          "kind": 23,
          "specifier": null,
          "start": 3,
          "type": "capturing",
        },
        {
          "charCode": 98,
          "end": 8,
          "kind": 6,
          "start": 8,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 2,
    },
    "end": 9,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "capturing",
  },
  "end": 10,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Negative lookahead > should parse simple negative lookahead 1`] = `
{
  "body": {
    "body": {
      "end": 5,
      "expressions": [
        {
          "charCode": 97,
          "end": 4,
          "kind": 6,
          "start": 4,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 5,
          "kind": 6,
          "start": 5,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 4,
    },
    "end": 6,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "negativeLookahead",
  },
  "end": 7,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Negative lookbehind > should parse simple negative lookbehind 1`] = `
{
  "body": {
    "body": {
      "end": 6,
      "expressions": [
        {
          "charCode": 97,
          "end": 5,
          "kind": 6,
          "start": 5,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 6,
          "kind": 6,
          "start": 6,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 5,
    },
    "end": 7,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "negativeLookbehind",
  },
  "end": 8,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Non Capturing group > should parse simple non capturing group 1`] = `
{
  "body": {
    "body": {
      "end": 5,
      "expressions": [
        {
          "charCode": 97,
          "end": 4,
          "kind": 6,
          "start": 4,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 5,
          "kind": 6,
          "start": 5,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 4,
    },
    "end": 6,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "nonCapturing",
  },
  "end": 7,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Positive lookahead > should parse simple positive lookahead 1`] = `
{
  "body": {
    "body": {
      "end": 5,
      "expressions": [
        {
          "charCode": 97,
          "end": 4,
          "kind": 6,
          "start": 4,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 5,
          "kind": 6,
          "start": 5,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 4,
    },
    "end": 6,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "positiveLookahead",
  },
  "end": 7,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Group > Positive lookbehind > should parse simple positive lookbehind 1`] = `
{
  "body": {
    "body": {
      "end": 6,
      "expressions": [
        {
          "charCode": 97,
          "end": 5,
          "kind": 6,
          "start": 5,
          "type": "simple",
          "value": "a",
        },
        {
          "charCode": 98,
          "end": 6,
          "kind": 6,
          "start": 6,
          "type": "simple",
          "value": "b",
        },
      ],
      "kind": 5,
      "start": 5,
    },
    "end": 7,
    "kind": 23,
    "specifier": null,
    "start": 1,
    "type": "positiveLookbehind",
  },
  "end": 8,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;
